### ISSUE TITLE
Qt Ctrl-C broken on windows

### ISSUE THREAD

**Issue by tacaswell** [matplotlib/matplotlib#20932]
@anntzer @tacaswell 
Hi! I have just tested the master branch on Windows and found that there is indeed a problem with the QSocketNotifier (it appeared here: https://github.com/matplotlib/matplotlib/pull/13306#issuecomment-760878434). Previously, on that Windows machine, I was running my old code with QAbstractSocket which was fine, and did not test the newest version with QSocketNotifier there (my fault, sorry). I have tested on Linux, and there is no such problem there.

The code to reproduce on Windows (in Jupyter):
```python
%matplotlib qt
from matplotlib import pyplot as plt
plt.plot([1,2])
plt.pause(.1)
plt.pause(.1)
plt.pause(.1)
```
The problem seems to be that if `socketpair` generates a socket with the same `fileno` value as it generated in the previous `pause()` call, the QSocketNotifier _unexpectedly_ fires and hangs here: https://github.com/matplotlib/matplotlib/blob/b600026fe86bf0c7e14fc56f568b8b1fae07313d/lib/matplotlib/backends/qt_compat.py#L232
because _nothing was actually written_ to the `wakeup_fd`. I don't understand how that can be.

Workaround fix:
```python
rsock.set_blocking(False)

@sn.activated.connect
def on_signal(*args):
    try:
        signal = rsock.recv(1)  # clear the socket to re-arm the notifier
    except BlockingIOError:
        pass # false-trigger, nothing on rsock
```

What should we do with this? Is there a better solution?

_Originally posted by @vdrhtc in https://github.com/matplotlib/matplotlib/issues/13306#issuecomment-906467433_

**Comment by QuLogic** [comment]
Is it the current `QSocketNotifier` that fires, or the previous one? Are we just missing some cleanup?

**Comment by vdrhtc** [comment]
@QuLogic 
![image](https://user-images.githubusercontent.com/3819012/131479015-e5ffca6a-f254-4408-9efe-87250f18f074.png)


Here how it looks like, I don't think it is the same QSocketNotifier, it is a different object every time. But false-trigger only happens if the `fileno` is re-used. So maybe it is necessary to clear the rsock somehow.

**Comment by QuLogic** [comment]
Does something like this work?
```python
import select
import socket


for i in range(10):
    wsock, rsock = socket.socketpair()
    wsock.setblocking(False)
    rfd = rsock.fileno()
    wfd = wsock.fileno()
    print('rsock:', rfd, 'wsock:', wfd)
    ready = select.select([rfd], [], [], 2)
    print('ready:', ready)
    wsock.close()
    rsock.close()
```
It should print out that no sockets are ready. Windows support was only added in 3.5, but maybe there's a bug?

Maybe we should call `socket.shutdown` on the sockets to force them to be disabled, even though we do close them:

>    Shut down one or both halves of the connection. If how is SHUT_RD, further receives are disallowed. If how is SHUT_WR, further sends are disallowed. If how is SHUT_RDWR, further sends and receives are disallowed.

**Comment by QuLogic** [comment]
Also, does reusing the existing socket pair work?

```python
wsock, rsock = socket.socketpair()
wsock.setblocking(False)
sn = QtCore.QSocketNotifier(
    rsock.fileno(), _enum('QtCore.QSocketNotifier.Type').Read
)
sn.activated.connect(lambda *args: print('activated 1'))
del sn
sn = QtCore.QSocketNotifier(
    rsock.fileno(), _enum('QtCore.QSocketNotifier.Type').Read
)
sn.activated.connect(lambda *args: print('activated 2'))
```

**Comment by vdrhtc** [comment]
Nothing unexpected in the first test despite occasional re-uses: 
![image](https://user-images.githubusercontent.com/3819012/131629353-424b7802-bda9-4601-adb8-97b503626209.png)

Second test also works:
![image](https://user-images.githubusercontent.com/3819012/131630682-a7c727ac-fece-448a-b798-5eccfcebdc4c.png)

If I remove the `del sn` line, then there is a warning about multiple notifiers for the same socket in the terminal, and no activation happens on both notifiers.

Calling 
```python
wsock.shutdown(socket.SHUT_RDWR)
rsock.shutdown(socket.SHUT_RDWR)
wsock.close()
rsock.close()
``` 
does not help.

**Comment by QuLogic** [comment]
Actually, for that second test, maybe we should run the event loop in between?
```python
wsock, rsock = socket.socketpair()
wsock.setblocking(False)

sn = QtCore.QSocketNotifier(
    rsock.fileno(), _enum('QtCore.QSocketNotifier.Type').Read
)
sn.activated.connect(lambda *args: print('activated 1'))
# 5 seconds of event processing should be enough?
event_loop = QtCore.QEventLoop()
timer = QtCore.QTimer.singleShot(5000, event_loop.quit)
qt_compat._exec(event_loop)

del sn

sn = QtCore.QSocketNotifier(
    rsock.fileno(), _enum('QtCore.QSocketNotifier.Type').Read
)
sn.activated.connect(lambda *args: print('activated 2'))
# 5 seconds of event processing should be enough?
event_loop = QtCore.QEventLoop()
timer = QtCore.QTimer.singleShot(5000, event_loop.quit)
qt_compat._exec(event_loop)
```

**Comment by vdrhtc** [comment]
@QuLogic For some reason, the event loop does not exit after 5s and then the program hangs...

I've obtained the previous results by running %matplotlib qt first (it probably starts the event loop correctly by itself)

**Comment by timhoffm** [comment]
Is this fixed by #20907?

**Comment by QuLogic** [comment]
It should be.
